src/hir/combiner.rs

```diff
@@ -0,0 +1,90 @@
+//! The implementation of a high-level intermediate representation (HIR) combiner.
+
+use std::collections::HashSet;
+
+use crate::syntax::ast::Ast;
+
+use super::{Hir, Root};
+
+/// A high-level intermediate representation (HIR) combiner.
+///
+/// It takes a vector of HIRs and combines them into a single HIR
+/// by appending the function nodes to the root contract node.
+pub struct Combiner {}
+
+impl Combiner {
+    /// Creates a new combiner.
+    #[must_use]
+    pub fn new() -> Self {
+        Combiner {}
+    }
+
+    /// Splits the input text into distinct trees,
+    /// delimited by two successive newlines
+    ///
+    /// This function is called before the tokenization and parsing steps.
+    pub fn split<'a>(&'a self, text: &'a str) -> Vec<&str> {
+        text.split("\n\n").collect::<Vec<&str>>()
+    }
+
+    /// Verifies that the ASTs are valid.
+    ///
+    /// This function is called before the translation to HIR.
+    pub fn verify(&self, asts: &Vec<Ast>) -> anyhow::Result<(), String> {
+        if let Some(Ast::Root(first_root)) = asts.first() {
+            let first_contract_name = &first_root.contract_name;
+            let first_parts: Vec<&str> = first_contract_name
+                .split(|c| c == '.' || c == ':')
+                .collect();
+
+            for ast in asts {
+                if let Ast::Root(root) = ast {
+                    let parts: Vec<&str> =
+                        root.contract_name.split(|c| c == '.' || c == ':').collect();
+                    if parts[0] != first_parts[0] {
+                        return Err(format!(
+                            "Contract name mismatch: expected '{}', found '{}'",
+                            first_contract_name, root.contract_name
+                        ));
```

Comment:
@alexfertel, I believe I have made some progress on this front, although I'm not quite sure what additional information we want to report and how. For example, I am having some difficulty working out how we can get the span of the original text from the HIR (more context on L125 of `combiner.rs`). Some input here would be great!

Reply:
>  I am having some difficulty working out how we can get the span of the original text from the HIR

Ah, yes, this is a limitation of the current design, since a HIR doesn't really map to a tree, this is something to figure out. No worries, we don't have to tackle this here since it is unrelated and uphill.